MVP Memo Table (DO NOT MERGE)

connortsui20 opened the pull request:
~~Very incomplete~~

Less incomplete

TODO:

- [ ] **We should stabilize the on-disk representation of expressions (kind + serializable and deserializable data, where each operator kind needs to know how to retrieve their children from this data field). Once this is stabilized, it should not change.**
- [ ] Potentially make APIs clearer. There are a bunch of `Result<Result<_, _>>` which can be unclear, but I have no idea how to make this better.
- [ ] Potentially should make a distinction at the type level between a `GroupId` and a `RootGroupId`, where `impl From<RootGroupId> for GroupId` is implemented but not the other way around...
- [ ] Since fingerprints are in a separate table, it might be good to do updates instead of additional inserts because the old fingerprints will all be stale and there shouldn't be that much contention on them?
- [ ] Should the union-find pointer data structure be separated out into its own table? Right now it is embedded into the `group` table records, but if we separate it out into a separate table we will have a level of abstraction that allows us to change the implementation when we need to.
- [x] Add union by size optimization after deciding if the union find data structure should be separated from the main group records.
- [ ] Logical expression semantics should be formalized and a trait should be made that represents these semantics.
- [x] All methods on the memo table should be serializable transactions, and the implementations should reflect that. This might mean that the transaction object needs to be passed around to different function calls, as the nested [transaction API](https://www.sea-ql.org/SeaORM/docs/advanced-query/transaction/) SeaORM might not be ideal (but maybe it is, we'll have to see).
- [ ] Add much more tests
- [ ] Add support for running tests in CI (should test both in-memory and disk based DBMS)
- [ ] Add code coverage in CI
- [ ] Add `tracing` and `log` support

connortsui20 committed:
add boilerplate files

This commit adds the boilerplate files and fixes the CI workflows.

connortsui20 committed:
add migration and entity files

connortsui20 committed:
add memo trait interface and persistent memo implementation

This commit adds a first draft of a memo table trait and a persistent
memo table implementation backed by SeaORM entities.

connortsui20 committed:
add expression representation and refactor memo

This commit adds the `src/expression` module which contains a very
simple representation of Cascades expressions.

The `Memo` trait interface and implemenation has also changed, where
it now correctly detects exact match duplicates, and it does not track
fingerprints for physical expressions (only logical).

TODO: Add more tests.
TODO: Figure out how to test in CI.

connortsui20 commented on .gitignore:
```diff
@@ -0,0 +1,13 @@
+
+# We will check in all code-generated entity files, as newer versions of `sea-orm-cli` might
+# conflict with previous versions.
+# **/entities
```
Missing newline

connortsui20 commented on optd-mvp/src/entities/prelude.rs:
```diff
@@ -0,0 +1,9 @@
+//! `SeaORM` Entity, @generated by sea-orm-codegen 1.1.0
+
+#![allow(unused_imports)]
```
Should eventually remove this

connortsui20 commented on optd-mvp/src/migrator/memo/m20241127_000001_cascades_group.rs:
```diff
@@ -0,0 +1,123 @@
+    Id,
+    Winner,
+    Cost,
+    IsOptimized,
```
`IsOptimized` should be an enum representing the different states a group could be in:

- No progress
- Exploring
- Fully Explored
- Optimizing
- Optimized

May or may not need the exploring and optimizing states, but it would be good to make this an enum regardless for extensibility and robustness.

See: https://www.sea-ql.org/SeaORM/docs/migration/writing-migration/

connortsui20 commented on optd-mvp/src/memo/interface.rs:
```diff
@@ -0,0 +1,155 @@
+        group_id: Self::GroupId,
+        physical_expression_id: Self::PhysicalExpressionId,
+    ) -> OptimizerResult<Option<Self::PhysicalExpressionId>>;
+
```
Should probably add another method that allows checking for a duplicate logical expression.

connortsui20 committed:
huge refactor of persistent memo implementation

This commit completely refactors the memo table, removing the `Memo`
trait and instead placing all methods directly on the `PersistentMemo`
structure itself.

This also cleans up some code in other places.

connortsui20 committed:
add union find and group merging support

connortsui20 committed:
rename cascades_group to group

connortsui20 committed:
first draft fully merge group

connortsui20 commented:
To run tests, use:

```sh
cargo t -- --test-threads 1 --ignored
```

They are ignored because tests don't work in CI. I need to figure out how to run migrations in CI before those will work.

Also, each of the tests assumes they are isolated. I need to convert all of the implementation methods to use transaction instead in order to make this thread safe, which won't be hard but will be a bit tedious.

skyzh commented:
You can create a sqlite in memory database and run the migrate function on it for every test case?

connortsui20 commented:
Yes itll be easy I just haven't had time to do it yet and it's not that important

connortsui20 committed:
remove top-level Expression type and rename to Default<>Expression

connortsui20 committed:
make memo table genric of expression traits

This commit replaces the specific expression types with traits that
define the behavior the in-memory represenations of both logical and
physical expressions need to have. Right now, the `PhysicalExpression`
trait does not do that much, but the `LogicalExpression` trait is super
important to how the persistent memo table works.

connortsui20 committed:
add errors and panics documentation in memo module

connortsui20 committed:
implement transaction on all memo table operations

connortsui20 committed:
add rank by size optimization + bump sea-orm-cli

This commit adds the rank by size optimization into the embedded
union-find data structure of the group sets. It also bumps the version
number of `sea-orm-cli` to `1.1.2`.
